apiserver/facades/agent/secretsmanager/secrets.go

```diff
@@ -344,20 +344,93 @@ func (s *SecretsManagerAPI) GetSecretContentInfo(args params.GetSecretContentArg
 	return result, nil
 }
 
-func (s *SecretsManagerAPI) getSecretContent(arg params.GetSecretContentArg) (*secrets.ContentParams, error) {
-	uri, err := coresecrets.ParseURI(arg.URI)
+func (s *SecretsManagerAPI) getOwnerSecretMetadata(uri *coresecrets.URI, label string) (*coresecrets.SecretMetadata, error) {
+	if uri == nil {
+		var err error
+		uri, err = s.secretsBackend.GetURIBySecretLabel(label, s.authTag)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+
+	md, err := s.secretsBackend.GetSecret(uri)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	if !s.canRead(uri, s.authTag) {
-		return nil, apiservererrors.ErrPerm
+	if s.authTag.String() != md.OwnerTag {
+		// Invisible for non owner users.
+		return nil, errors.NotFoundf("secret %s, label %q", uri, label)
+	}
+	return md, nil
+}
+
+func (s *SecretsManagerAPI) getSecretContent(arg params.GetSecretContentArg) (*secrets.ContentParams, error) {
+	// Only the owner can access secrets via the secret label (Note: the leader unit is not the owner of the application secrets).
+	// Non owners have to use secret URI or consumer label.
+
+	if arg.URI == "" && arg.Label == "" {
+		return nil, errors.NewNotValid(nil, "both uri and label are empty")
 	}
+
+	getSecretValue := func(uri *coresecrets.URI, revision int) (*secrets.ContentParams, error) {
+		if !s.canRead(uri, s.authTag) {
+			return nil, apiservererrors.ErrPerm
+		}
+
+		val, providerId, err := s.secretsBackend.GetSecretValue(uri, revision)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		return &secrets.ContentParams{SecretValue: val, ProviderId: providerId}, nil
+	}
+
+	var uri *coresecrets.URI
+	var err error
+
+	if arg.URI != "" {
+		uri, err = coresecrets.ParseURI(arg.URI)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+
+	// arg.Label is the secret label for owner.
```

Comment:
arg.Label could also be the consumer label too
The comment here should be along the lines of we will first check to see if the label corresponds to a secret owned by the caller.